// stylelint/dist/rules/selector-pseudo-class-no-unknown/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual, options) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual }, {
      actual: options,
      possible: {
        ignorePseudoClasses: [_lodash.isString]
      },
      optional: true
    });
    if (!validOptions) {
      return;
    }

    root.walkRules(function (rule) {
      if (!(0, _utils.isStandardSyntaxRule)(rule)) {
        return;
      }
      var selector = rule.selector;

      // Return early before parse if no pseudos for performance

      if (selector.indexOf(":") === -1) {
        return;
      }

      (0, _utils.parseSelector)(selector, result, rule, function (selectorTree) {
        selectorTree.walkPseudos(function (pseudoNode) {
          var value = pseudoNode.value;


          if (!(0, _utils.isStandardSyntaxSelector)(value)) {
            return;
          }

          // Ignore pseudo-elements
          if (value.slice(0, 2) === "::") {
            return;
          }

          if ((0, _utils.optionsMatches)(options, "ignorePseudoClasses", pseudoNode.value.slice(1))) {
            return;
          }

          var name = value.slice(1);

          if (_postcss.vendor.prefix(name) || _keywordSets.pseudoClasses.has(name.toLowerCase()) || _keywordSets.pseudoElements.has(name.toLowerCase())) {
            return;
          }

          (0, _utils.report)({
            message: messages.rejected(value),
            node: rule,
            index: pseudoNode.sourceIndex,
            ruleName: ruleName,
            result: result
          });
        });
      });
    });
  };
};

var _utils = require("../../utils");

var _keywordSets = require("../../reference/keywordSets");

var _lodash = require("lodash");

var _postcss = require("postcss");

var ruleName = exports.ruleName = "selector-pseudo-class-no-unknown";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(selector) {
    return "Unexpected unknown pseudo-class selector \"" + selector + "\"";
  }
});
