// stylelint/dist/rules/custom-property-empty-line-before/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (expectation, options) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: expectation,
      possible: ["always", "never"]
    }, {
      actual: options,
      possible: {
        except: ["first-nested", "after-comment", "after-custom-property"],
        ignore: ["after-comment", "inside-single-line-block"]
      },
      optional: true
    });
    if (!validOptions) {
      return;
    }

    root.walkDecls(function (decl) {
      var prop = decl.prop,
          parent = decl.parent;


      if (!(0, _utils.isStandardSyntaxDeclaration)(decl)) {
        return;
      }
      if (!(0, _utils.isCustomProperty)(prop)) {
        return;
      }

      // Optionally ignore the node if a comment precedes it
      if ((0, _utils.optionsMatches)(options, "ignore", "after-comment") && decl.prev() && decl.prev().type === "comment") {
        return;
      }

      // Optionally ignore nodes inside single-line blocks
      if ((0, _utils.optionsMatches)(options, "ignore", "inside-single-line-block") && (0, _utils.isSingleLineString)((0, _utils.blockString)(parent))) {
        return;
      }

      var expectEmptyLineBefore = expectation === "always" ? true : false;

      // Optionally reverse the expectation for the first nested node
      if ((0, _utils.optionsMatches)(options, "except", "first-nested") && decl === parent.first) {
        expectEmptyLineBefore = !expectEmptyLineBefore;
      }

      // Optionally reverse the expectation if a comment precedes this node
      if ((0, _utils.optionsMatches)(options, "except", "after-comment") && decl.prev() && decl.prev().type === "comment") {
        expectEmptyLineBefore = !expectEmptyLineBefore;
      }

      // Optionally reverse the expectation if a custom property precedes this node
      if ((0, _utils.optionsMatches)(options, "except", "after-custom-property") && decl.prev() && decl.prev().prop && (0, _utils.isCustomProperty)(decl.prev().prop)) {
        expectEmptyLineBefore = !expectEmptyLineBefore;
      }

      var hasEmptyLineBefore = (0, _utils.hasEmptyLine)(decl.raws["before"]);

      // Return if the expectation is met
      if (expectEmptyLineBefore === hasEmptyLineBefore) {
        return;
      }

      var message = expectEmptyLineBefore ? messages.expected : messages.rejected;
      (0, _utils.report)({
        message: message,
        node: decl,
        result: result,
        ruleName: ruleName
      });
    });
  };
};

var _utils = require("../../utils");

var ruleName = exports.ruleName = "custom-property-empty-line-before";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  expected: "Expected empty line before custom property",
  rejected: "Unexpected empty line before custom property"
});
