// stylelint/dist/rules/media-feature-name-no-unknown/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual, options) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual }, {
      actual: options,
      possible: {
        ignoreMediaFeatureNames: [_lodash.isString]
      },
      optional: true
    });

    if (!validOptions) {
      return;
    }

    root.walkAtRules(/^media$/i, function (atRule) {
      (0, _postcssMediaQueryParser2.default)(atRule.params).walk(/^media-feature$/i, function (mediaFeatureNode) {
        var parent = mediaFeatureNode.parent,
            sourceIndex = mediaFeatureNode.sourceIndex,
            value = mediaFeatureNode.value;


        if ((0, _utils.isRangeContextMediaFeature)(parent.value) || !(0, _utils.isStandardSyntaxMediaFeatureName)(value) || (0, _utils.isCustomMediaQuery)(value)) {
          return;
        }

        if ((0, _utils.optionsMatches)(options, "ignoreMediaFeatureNames", value)) {
          return;
        }

        if (_postcss.vendor.prefix(value) || _keywordSets.mediaFeatureNames.has(value.toLowerCase())) {
          return;
        }

        (0, _utils.report)({
          index: (0, _utils.atRuleParamIndex)(atRule) + sourceIndex,
          message: messages.rejected(value),
          node: atRule,
          ruleName: ruleName,
          result: result
        });
      });
    });
  };
};

var _utils = require("../../utils");

var _lodash = require("lodash");

var _keywordSets = require("../../reference/keywordSets");

var _postcssMediaQueryParser = require("postcss-media-query-parser");

var _postcssMediaQueryParser2 = _interopRequireDefault(_postcssMediaQueryParser);

var _postcss = require("postcss");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ruleName = exports.ruleName = "media-feature-name-no-unknown";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(mediaFeatureName) {
    return "Unexpected unknown media feature name \"" + mediaFeatureName + "\"";
  }
});
