// stylelint/dist/rules/function-url-no-scheme-relative/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual });
    if (!validOptions) {
      return;
    }

    root.walkDecls(function (decl) {
      (0, _utils.functionArgumentsSearch)(decl.toString().toLowerCase(), "url", function (args, index) {
        var url = (0, _lodash.trim)(args, " '\"");

        if (!(0, _utils.isStandardSyntaxUrl)(url) || url.indexOf("//") !== 0) {
          return;
        }

        (0, _utils.report)({
          message: messages.rejected,
          node: decl,
          index: index,
          result: result,
          ruleName: ruleName
        });
      });
    });
  };
};

var _utils = require("../../utils");

var _lodash = require("lodash");

var ruleName = exports.ruleName = "function-url-no-scheme-relative";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: "Unexpected scheme-relative url"
});
