// stylelint/dist/rules/selector-descendant-combinator-no-non-space/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual });
    if (!validOptions) {
      return;
    }

    root.walkRules(function (rule) {
      if (!(0, _utils.isStandardSyntaxRule)(rule)) {
        return;
      }

      var selector = rule.selector;


      (0, _utils.parseSelector)(selector, result, rule, function (fullSelector) {
        fullSelector.walkCombinators(function (combinatorNode) {
          var value = combinatorNode.value;


          if (_punctuationSets.nonSpaceCombinators.has(value)) {
            return;
          }
          if (value === " ") {
            return;
          }

          (0, _utils.report)({
            result: result,
            ruleName: ruleName,
            message: messages.rejected(value),
            node: rule,
            index: combinatorNode.sourceIndex
          });
        });
      });
    });
  };
};

var _utils = require("../../utils");

var _punctuationSets = require("../../reference/punctuationSets");

var ruleName = exports.ruleName = "selector-descendant-combinator-no-non-space";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(nonSpaceCharacter) {
    return "Unexpected \"" + nonSpaceCharacter + "\"";
  }
});
