// stylelint/dist/rules/selector-pseudo-class-whitelist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

var _utils = require("../../utils");

var _lodash = require("lodash");

var _postcss = require("postcss");

var ruleName = exports.ruleName = "selector-pseudo-class-whitelist";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(selector) {
    return "Unexpected pseudo-class \"" + selector + "\"";
  }
});

function rule(whitelist) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: whitelist,
      possible: [_lodash.isString]
    });
    if (!validOptions) {
      return;
    }

    root.walkRules(function (rule) {
      var selector = rule.selector;


      if (!(0, _utils.isStandardSyntaxSelector)(selector)) {
        return;
      }
      if (selector.indexOf(":") === -1) {
        return;
      }

      (0, _utils.parseSelector)(selector, result, rule, function (selectorTree) {
        selectorTree.walkPseudos(function (pseudoNode) {
          var value = pseudoNode.value;

          // Ignore pseudo-elements

          if (value.slice(0, 2) === "::") {
            return;
          }

          var name = value.slice(1);

          if ((0, _utils.matchesStringOrRegExp)(_postcss.vendor.unprefixed(name).toLowerCase(), whitelist)) {
            return;
          }

          (0, _utils.report)({
            index: pseudoNode.sourceIndex,
            message: messages.rejected(name),
            node: rule,
            result: result,
            ruleName: ruleName
          });
        });
      });
    });
  };
}

rule.primaryOptionArray = true;

exports.default = rule;
