// electron-builder/out/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _packager;

function _load_packager() {
  return _packager = require("./packager");
}

Object.defineProperty(exports, "Packager", {
  enumerable: true,
  get: function () {
    return (_packager || _load_packager()).Packager;
  }
});

var _targetFactory;

function _load_targetFactory() {
  return _targetFactory = require("./targets/targetFactory");
}

Object.defineProperty(exports, "DIR_TARGET", {
  enumerable: true,
  get: function () {
    return (_targetFactory || _load_targetFactory()).DIR_TARGET;
  }
});
Object.defineProperty(exports, "DEFAULT_TARGET", {
  enumerable: true,
  get: function () {
    return (_targetFactory || _load_targetFactory()).DEFAULT_TARGET;
  }
});

var _builder;

function _load_builder() {
  return _builder = require("./builder");
}

Object.defineProperty(exports, "build", {
  enumerable: true,
  get: function () {
    return (_builder || _load_builder()).build;
  }
});
Object.defineProperty(exports, "createTargets", {
  enumerable: true,
  get: function () {
    return (_builder || _load_builder()).createTargets;
  }
});

var _publisher;

function _load_publisher() {
  return _publisher = require("./publish/publisher");
}

Object.defineProperty(exports, "Publisher", {
  enumerable: true,
  get: function () {
    return (_publisher || _load_publisher()).Publisher;
  }
});

var _metadata;

function _load_metadata() {
  return _metadata = require("./metadata");
}

Object.defineProperty(exports, "Platform", {
  enumerable: true,
  get: function () {
    return (_metadata || _load_metadata()).Platform;
  }
});
Object.defineProperty(exports, "Arch", {
  enumerable: true,
  get: function () {
    return (_metadata || _load_metadata()).Arch;
  }
});
Object.defineProperty(exports, "archFromString", {
  enumerable: true,
  get: function () {
    return (_metadata || _load_metadata()).archFromString;
  }
});
