// stylelint/dist/rules/at-rule-no-unknown/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual, options) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual }, {
      actual: options,
      possible: {
        ignoreAtRules: [_lodash.isString]
      },
      optional: true
    });

    if (!validOptions) {
      return;
    }

    root.walkAtRules(function (atRule) {
      var name = atRule.name;

      // Return early if at-rule is to be ignored

      if ((0, _utils.optionsMatches)(options, "ignoreAtRules", atRule.name)) {
        return;
      }

      if (_postcss.vendor.prefix(name) || _keywordSets.atRules.has(name.toLowerCase())) {
        return;
      }

      (0, _utils.report)({
        message: messages.rejected("@" + name),
        node: atRule,
        ruleName: ruleName,
        result: result
      });
    });
  };
};

var _utils = require("../../utils");

var _keywordSets = require("../../reference/keywordSets");

var _lodash = require("lodash");

var _postcss = require("postcss");

var ruleName = exports.ruleName = "at-rule-no-unknown";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(atRule) {
    return "Unexpected unknown at-rule \"" + atRule + "\"";
  }
});
