// stylelint/dist/rules/media-feature-parentheses-space-inside/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (expectation) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: expectation,
      possible: ["always", "never"]
    });
    if (!validOptions) {
      return;
    }

    root.walkAtRules(/^media$/i, function (atRule) {
      // If there are comments in the params, the complete string
      // will be at atRule.raws.params.raw
      var params = _lodash2.default.get(atRule, "raws.params.raw", atRule.params);
      var indexBoost = (0, _utils.atRuleParamIndex)(atRule);

      (0, _styleSearch2.default)({ source: params, target: "(" }, function (match) {
        var nextCharIsSpace = params[match.startIndex + 1] === " ";
        if (nextCharIsSpace && expectation === "never") {
          (0, _utils.report)({
            message: messages.rejectedOpening,
            node: atRule,
            index: match.startIndex + 1 + indexBoost,
            result: result,
            ruleName: ruleName
          });
        }
        if (!nextCharIsSpace && expectation === "always") {
          (0, _utils.report)({
            message: messages.expectedOpening,
            node: atRule,
            index: match.startIndex + 1 + indexBoost,
            result: result,
            ruleName: ruleName
          });
        }
      });

      (0, _styleSearch2.default)({ source: params, target: ")" }, function (match) {
        var prevCharIsSpace = params[match.startIndex - 1] === " ";
        if (prevCharIsSpace && expectation === "never") {
          (0, _utils.report)({
            message: messages.rejectedClosing,
            node: atRule,
            index: match.startIndex - 1 + indexBoost,
            result: result,
            ruleName: ruleName
          });
        }
        if (!prevCharIsSpace && expectation === "always") {
          (0, _utils.report)({
            message: messages.expectedClosing,
            node: atRule,
            index: match.startIndex - 1 + indexBoost,
            result: result,
            ruleName: ruleName
          });
        }
      });
    });
  };
};

var _utils = require("../../utils");

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

var _styleSearch = require("style-search");

var _styleSearch2 = _interopRequireDefault(_styleSearch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ruleName = exports.ruleName = "media-feature-parentheses-space-inside";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  expectedOpening: "Expected single space after \"(\"",
  rejectedOpening: "Unexpected whitespace after \"(\"",
  expectedClosing: "Expected single space before \")\"",
  rejectedClosing: "Unexpected whitespace before \")\""
});
