// stylelint/dist/rules/comment-word-blacklist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

var _utils = require("../../utils");

var _lodash = require("lodash");

var ruleName = exports.ruleName = "comment-word-blacklist";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(pattern) {
    return "Unexpected word matching pattern \"" + pattern + "\"";
  }
});

function rule(blacklist) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: blacklist,
      possible: [_lodash.isString]
    });
    if (!validOptions) {
      return;
    }

    root.walkComments(function (comment) {
      var text = comment.text;
      var rawComment = comment.toString();
      var firstFourChars = rawComment.substr(0, 4);

      // Return early if sourcemap
      if (firstFourChars === "/*# ") {
        return;
      }

      var matchesWord = (0, _utils.matchesStringOrRegExp)(text, blacklist) || (0, _utils.containsString)(text, blacklist);

      if (!matchesWord) {
        return;
      }

      (0, _utils.report)({
        message: messages.rejected(matchesWord.pattern),
        node: comment,
        result: result,
        ruleName: ruleName
      });
    });
  };
}

rule.primaryOptionArray = true;

exports.default = rule;
