// stylelint/dist/rules/at-rule-name-newline-after/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (expectation) {
  var checker = (0, _utils.whitespaceChecker)("newline", expectation, messages);
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: expectation,
      possible: ["always", "always-multi-line"]
    });
    if (!validOptions) {
      return;
    }

    (0, _atRuleNameSpaceAfter.atRuleNameSpaceChecker)({
      root: root,
      result: result,
      locationChecker: checker.afterOneOnly,
      checkedRuleName: ruleName
    });
  };
};

var _utils = require("../../utils");

var _atRuleNameSpaceAfter = require("../at-rule-name-space-after");

var ruleName = exports.ruleName = "at-rule-name-newline-after";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  expectedAfter: function expectedAfter(name) {
    return "Expected newline after at-rule name \"" + name + "\"";
  }
});
